core/src/main/java/org/apache/iceberg/avro/ProjectionDatumReader.java

```diff
@@ -51,6 +52,10 @@ public ProjectionDatumReader(Function<Schema, DatumReader<?>> getReader,
   @Override
   public void setSchema(Schema newFileSchema) {
     this.fileSchema = newFileSchema;
+    boolean hasIds = AvroSchemaUtil.hasIds(fileSchema);
+    if (!hasIds && nameMapping == null) {
```

Comment:
Minor: it would be better to check if `nameMapping` is null first, and move the `hasIds` call into the condition. That way we avoid recursing over the schema if an explicit name mapping is going to be used instead.

Reply:
nice catch!